### ISSUE TITLE
Add option to Instrumentation CR to restart depoyments if config changes

### ISSUE THREAD

**Issue by pavolloffay** [open-telemetry/opentelemetry-operator#553]
I would like to have a way to trigger a rolling update if an instrumentation CR changes - e.g. the auto-instrumentation image is updated or sampling is changed.

**Comment by jpkrohling** [comment]
Good idea, but I believe this should be opt-in, at least in the first version of the CRD.

**Comment by pavolloffay** [comment]
exactly, that is what the title says :)

**Comment by Sprakhar97** [comment]
Also on a slightly different lines there should be a way to handle failed deployments caused by the mutation of pods by the instrumentation webhooks, there should be a mechanism to revert these changes or reapply the settings so that it doesnot affect the deployment of users application.

**Comment by frzifus** [comment]
Hey I'm working on this and would like some feedback on a draft/proposal.
So far I added a field `AutoUpdate` of type `bool` to the `InstrumentationSpec`. By default this would be disabled. 

Next, I introduced a reconciler to update corresponding pods when an `InstrumentationSpec` is created, deleted or modified. 
To verify that a pod configuration is up-to-date, I store the `sha256` sum of the configuration in an annotation. 

Originally I thought if the `sha256` sum of the `InstrumentationSpec` in the pod annotaion differs, I would force the restart by updating the annotation. Unfortunately, this does not work on the pod basis itself. My next thought was to realize it via `DaemonSet` and `Deployment`. However, this excludes differently created pods.
Do you have any idea how to solve the problem?

A solution could be to use the ownerID to search for the "highest" object that contains a pod specification and change the pod annotation there. However, this way seems to me to be a "dirty hack" that still would not work on standalone pods without an owner.

**Comment by pavolloffay** [comment]
Some notes from the meeting with @frzifus:
* support only daemonset, statefulset, deployment (are these all replicate set based objects?). For instance it does not make sense to restart cronjob, job. 
* gracefully restart the pod
* do we need to store SHA annotation at all? The injection should be idempotent and from the pod spec and instrumentation CR we should be able to recognize if something changed

**Comment by jaronoff97** [comment]
@frzifus is this done?

**Comment by frzifus** [comment]
No, but it should now be easier to implement then two years ago :facepalm:

**Comment by jaronoff97** [comment]
Would you be able to take a look at how hard this would be to do?
